daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/LayerDriver.scala

```diff
@@ -0,0 +1,338 @@
+import passera.unsigned.ULong
+
+/**
+ * Mandatory factory to create and initialize the layer driver object.
+ */
+object LayerDriver {
+  def apply(state: ParseOrUnparseState, lrd: LayerRuntimeData): LayerDriver = {
+    val lvr = state.layerRuntimeCompiler.compile(lrd) // could fail if layer not SPI loaded
+    val lr = new LayerRuntime(state, lrd, lvr) // can't fail
+    val newLayer = lvr.constructInstance() // can't fail
+    newLayer.setLayerRuntime(lr) // can't fail
+    val instance = new LayerDriver(newLayer) // can't fail
+    lvr.callParamSetterWithParameterVars(newLayer) // could fail if user's method causes errors
+    instance
+  }
+}
+
+/**
+ * Driving mechanism that incorporates a layer at runtime into the I/O streams
+ * to transform the data stream and optionally receive parameter values from,
+ * and populate result values back into DFDL variables.
+ *
+ * A layer driver is created at runtime as part of a single parse/unparse call.
+ * Hence, they can be stateful without causing thread-safety issues.
+ *
+ * @param layer            The Layer instance to parameterize, drive, and get results back from.
+ */
+class LayerDriver private (val layer: Layer) {
+
+  def layerRuntime: LayerRuntime = layer.getLayerRuntime
+  private def layerVarsRuntime = layerRuntime.layerVarsRuntime
+
+  private def wrapJavaInputStream(s: InputSourceDataInputStream) =
+    new JavaIOInputStream(s, layer)
+
+  private def wrapJavaOutputStream(s: DataOutputStream) =
+    new JavaIOOutputStream(s, layer)
+
+  /**
+   *  Override these if we ever have transformers that don't have these
+   *  requirements.
+   */
+  val mandatoryLayerAlignmentInBits: Int = 8
+
+  final def addInputLayer(s: InputSourceDataInputStream): InputSourceDataInputStream = {
+    val jis = wrapJavaInputStream(s)
+    // This wrapLayerInput could fail via a throw. It's calling layer code.
+    val wrapped =
+      try {
+        layer.wrapLayerInput(jis)
+      } catch {
+        case t: Throwable => escalateThrowFromWrap(t)
+      }
+    // This assured close thing is just in case the user's layer stream doesn't propagate
+    // the close. That would be incorrect of it, but we want to tolerate that not happening.
+    val decodedInputStream = new AssuredCloseInputStream(wrapped, jis)
+    val newDIS = InputSourceDataInputStream(decodedInputStream)
+    // must initialize priorBitOrder
+    newDIS.cst.setPriorBitOrder(BitOrder.MostSignificantBitFirst)
+    newDIS.setDebugging(s.areDebugging)
+    newDIS
+  }
+
+  /**
+   * Parsing works as a tree traversal, so when the parser unwinds from
+   * parsing the layer we can invoke this to handle cleanups, and
+   * finalization issues like assigning the result variables
+   */
+  final def removeInputLayer(s: InputSourceDataInputStream): Unit =
+    try {
+      layerVarsRuntime.callGettersToPopulateResultVars(layer) // could throw.
+    } catch {
+      case ite: InvocationTargetException =>
+        layer.processingError(ite.getCause)
```

Comment:
No coverage here is surprisng to me. There are definitely tests that throw from within the special getter method, which is called reflectively, so should hit this catch.